File: core/x/javaapi/api/src/main/java/viaduct/java/api/types/MultidirectionalConnectionArguments.java
package viaduct.java.api.types;

import viaduct.tenant.runtime.support.ConnectionArgumentsSupport;

/**
 * Arguments for connections that expose all four pagination args ({@code first}, {@code after},
 * {@code last}, {@code before}).
 *
 * <p>Forward args take precedence over backward in {@link #toOffsetLimit(int)}; passing neither
 * returns the first page at the default size. Forward and backward args cannot be mixed in the same
 * request.
 *
 * <p>Java equivalent of Kotlin's {@code viaduct.api.types.MultidirectionalConnectionArguments}. The
 * direction dispatch, math, and validation are delegated to the language-neutral {@link
 * ConnectionArgumentsSupport} shared with the Kotlin Tenant API (which validates exactly once).
 *
 * <p>Prefer {@link ForwardConnectionArguments} or {@link BackwardConnectionArguments} when only one
 * direction is needed.
 */
public interface MultidirectionalConnectionArguments
    extends ForwardConnectionArguments, BackwardConnectionArguments {

  /**
   * Returns true if backward pagination is active ({@code last}/{@code before}) and requires total
   * count. Forward pagination never requires total count; backward pagination requires it only when
   * the {@code before} cursor is absent.
   */
  @Override
  default boolean requiresTotalCountForOffsetLimit() {
    return ConnectionArgumentsSupport.INSTANCE.multidirectionalRequiresTotalCount(
        getFirst(), getAfter(), getLast(), getBefore());
  }

  /**
   * Converts multidirectional pagination arguments to offset/limit. Uses forward pagination ({@code
   * first}/{@code after}) if provided, otherwise falls back to backward pagination ({@code
   * last}/{@code before}).
   */
  @Override
  default OffsetLimit toOffsetLimit(int defaultPageSize) {
    ConnectionArgumentsSupport.OffsetBounds b =
        ConnectionArgumentsSupport.INSTANCE.multidirectionalOffsetLimit(
            getFirst(), getAfter(), getLast(), getBefore(), defaultPageSize);
    return new OffsetLimit(b.getOffset(), b.getLimit());
  }

  /**
   * Converts multidirectional pagination arguments to offset/limit when total count is known. Uses
   * forward pagination ({@code first}/{@code after}) if provided, otherwise falls back to backward
   * pagination ({@code last}/{@code before}).
   */
  @Override
  default OffsetLimit toOffsetLimit(int totalCount, int defaultPageSize) {
    ConnectionArgumentsSupport.OffsetBounds b =
        ConnectionArgumentsSupport.INSTANCE.multidirectionalOffsetLimit(
            getFirst(), getAfter(), getLast(), getBefore(), totalCount, defaultPageSize);
    return new OffsetLimit(b.getOffset(), b.getLimit());
  }

  /**
   * Validates multidirectional pagination arguments.
   *
   * @throws IllegalArgumentException if mixing forward and backward pagination, or if individual
   *     arguments are invalid
   */
  @Override
  default void validate() {
    ConnectionArgumentsSupport.INSTANCE.validateMultidirectional(
        getFirst(), getAfter(), getLast(), getBefore());
  }
}

File: core/tenant/api/src/main/kotlin/viaduct/api/types/MultidirectionalConnectionArguments.kt
package viaduct.api.types

import viaduct.apiannotations.ExperimentalApi
import viaduct.tenant.runtime.support.ConnectionArgumentsSupport

/**
 * Arguments for connections that expose all four pagination args (`[first]`, `[after]`, `[last]`, `[before]`).
 *
 * Forward args take precedence over backward in [toOffsetLimit]; passing neither returns the first
 * page at the default size. Forward and backward args cannot be mixed in the same request.
 *
 * The math and validation are delegated to the language-neutral [ConnectionArgumentsSupport], which
 * validates exactly once per call (the previous inheritance-based implementation validated twice).
 *
 * Prefer [ForwardConnectionArguments] or [BackwardConnectionArguments] when only one direction
 * is needed.
 */
@ExperimentalApi
interface MultidirectionalConnectionArguments :
    ForwardConnectionArguments, BackwardConnectionArguments {
    /**
     * Returns true if backward pagination is active ([last]/[before]) and requires total count.
     *
     * Forward pagination never requires total count. Backward pagination requires it only when the
     * [before] cursor is absent.
     */
    @ExperimentalApi
    override fun requiresTotalCountForOffsetLimit(): Boolean = ConnectionArgumentsSupport.multidirectionalRequiresTotalCount(first, after, last, before)

    /**
     * Converts multidirectional pagination arguments to offset/limit.
     *
     * Uses forward pagination ([first]/[after]) if provided, otherwise falls back to backward
     * pagination ([last]/[before]).
     */
    @ExperimentalApi
    override fun toOffsetLimit(defaultPageSize: Int): OffsetLimit =
        ConnectionArgumentsSupport.multidirectionalOffsetLimit(first, after, last, before, defaultPageSize).let {
            OffsetLimit(offset = it.offset, limit = it.limit)
        }

    /**
     * Converts multidirectional pagination arguments to offset/limit when total count is known.
     *
     * Uses forward pagination ([first]/[after]) if provided, otherwise falls back to backward
     * pagination ([last]/[before]).
     */
    @ExperimentalApi
    override fun toOffsetLimit(
        totalCount: Int,
        defaultPageSize: Int
    ): OffsetLimit =
        ConnectionArgumentsSupport.multidirectionalOffsetLimit(first, after, last, before, totalCount, defaultPageSize).let {
            OffsetLimit(offset = it.offset, limit = it.limit)
        }

    /**
     * Validates multidirectional pagination arguments.
     *
     * @throws IllegalArgumentException if mixing forward and backward pagination, or if individual
     *     arguments are invalid
     */
    @ExperimentalApi
    override fun validate() = ConnectionArgumentsSupport.validateMultidirectional(first, after, last, before)
}
